### PR TITLE
Adding implicit-str-concat-in-sequence check

### PR DESCRIPTION
**Pull request by Lucas-C**
### Fixes / new features
- Adds a new warning. Example of code that would raise it:
```
TEST_LIST = ('a', 'b' 'c')
```
Message:
```
String literals are implicitly concatenated in a comma-separated list : maybe a comma is missing ?
```

### CODE REVIEW

**Comment by rogalski** [comment]
Hello,

Thank you for you PR!

Can we have some test case along lines of:
```python
def my_func(arg, arg2, msg):
    pass

my_func("s1", "s2", "really long message that I'd like to "
                    "split across multiple lines")
```

Case above (IMHO) shouldn't emit a warning but it looks like it would be emitted.

**Comment by Lucas-C** [comment]
Well spotted, thanks
I'm going to add it and fix the checker

**Comment by Lucas-C** [comment]
Ok, so i'm facing a dead-end here :(

First, this problem has a bit of history:
- https://github.com/PyCQA/pylint/issues/1589
- https://stackoverflow.com/questions/34540634/issue-warning-for-missing-comma-between-list-items-bug

Now, I see 2 approaches in `pylint`:

- use `astroid` AST visitor-pattern. Sadly it is limited by Python builtin `compile` method (cf. [usage in code](https://github.com/PyCQA/astroid/blob/master/astroid/builder.py#L41)) which does not allow to detect string concatenation : at its level, it's just a string node

- analyze tokens generated with `tokenize`. This time, the problem is that I cannot detect if I'm in a `list`/`tuple` context, or if it's function arguments as in your example

The only idea I've got would be to retrieve in the source code the code snippet corresponding to each `list`/`tuple` element, to check if there is string concatenation.

Did any checker had to implement such thing already ? I.e retrieving source code from `node.lineno` / `node.col_offset` ?

**Comment by asottile** [comment]
I'm not sure if pylint gives you an option to but one way would be to cross reference the ast nodes with the tokenized output -- this is roughly how [pyupgrade](https://github.com/asottile/pyupgrade) finds list / tuple nodes.

Otherwise disregard me, just drive by browsing issues (while I wait for pylint to clone) and noticed a familiar name :)

**Comment by Lucas-C** [comment]
Hey, thanks for the advice @asottile !

I contemplated this idea, but it's awesome to see you already implemented it quite concisely:
https://github.com/asottile/pyupgrade/blob/master/pyupgrade.py#L370-L383

Before implementing something like that in `pylint` though, it'd be great to know:
- is there already something like that in `pylint` ? I.e cross-referencing tokens & ast
- else, in which module would it be best to put this ?  `pylint/utils.py` ?

@PCManticore, @rogalski ?

**Comment by rogalski** [comment]
@Lucas-C You'll need a checker that inherits from `BaseTokenChecker`. From what I see in code, processing tokens occurs before processing AST nodes, so potential implementation may look like:

1. Collect all "suspiciously" joined strings in process_tokens (which is already implemented), store information about them on some class attribute
2. For each const node visited:
3. If constant node position (row / col) matches suspicious string found earlier - emit a message

**Comment by Lucas-C** [comment]
Thanks @rogalski  !

I updated the PR, I think I nailed it ;)

File: ChangeLog [hunk 1]
```diff
@@ -102,6 +102,10 @@ What's New in Pylint 1.8?
 
       Close #574
 
+    * Check for string concatenation inside a list, introducing new
```
Comment (rogalski) [inline ChangeLog]: It's a lie, you do not check list, but also tuples.

What about set literals? To be consistent, I'd expect to see lists, tuples and sets supported

File: ChangeLog [hunk 2]
```diff
@@ -102,6 +102,10 @@ What's New in Pylint 1.8?
...
+      W1403 message (patch by Lucas Cimon)
```
Comment (rogalski) [inline ChangeLog]: I'd really prefer to see `message_name` instead of `Wxxxx`

File: pylint/checkers/strings.py [hunk 3]
```diff
@@ -104,7 +105,11 @@
               "Used when a PEP 3101 format string uses a lookup specifier "
               "({a[1]}), but the argument passed for formatting "
               "doesn't contain or doesn't have that key as an attribute.",
-              {'minversion': (2, 7)})
+              {'minversion': (2, 7)}),
+    'W1308': ("Mixing string literals concatenation with comma-separated %s",
+              "str-literals-concatenation-and-comma",
+              "String literals are implicitly concatenated in a "
+              "comma-separated list : maybe a comma is missing ?")
```
Comment (rogalski) [inline pylint/checkers/strings.py]: _comma-separated **list**_ lies when message is emitted for tuple

File: pylint/checkers/strings.py [hunk 4]
```diff
@@ -234,6 +239,22 @@ class StringFormatChecker(BaseChecker):
...
+
+    def check_for_concatenated_strings(self, iterable_node, iterable_type):
+        # This check relies on the asumption that StringConstantChecker is run beforehand
+        for elt in iterable_node.elts:
+            if isinstance(elt, astroid.node_classes.Const) and elt.pytype() == 'builtins.str':
+                if StringConstantChecker.string_tokens[(elt.lineno, elt.col_offset)] != elt.value:
+                    self.add_message('str-literals-concatenation-and-comma',
+                                     node=iterable_node, args=(iterable_type,))
+
+    @check_messages(*(set(MSGS.keys()) ))
```
Comment (rogalski) [inline pylint/checkers/strings.py]: creating set from dict keys is redundant.

File: pylint/test/functional/str_literals_concatenation_and_comma.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
+#pylint: disable=bad-continuation,invalid-name,missing-docstring
```
Comment (rogalski) [inline pylint/test/functional/str_literals_concatenation_and_comma.py]: your tests does not pass.

File: pylint/checkers/strings.py [hunk 6]
```diff
@@ -541,15 +562,21 @@ class StringConstantChecker(BaseTokenChecker):
     # Unicode strings.
     UNICODE_ESCAPE_CHARACTERS = 'uUN'
 
+    string_tokens = {}
```
Comment (rogalski) [inline pylint/checkers/strings.py]: this really should be instance attribute, no class attribute.
We also need to be sure that data gets thrown away before switching to new module for analysis (simple reinitialization of dict on the beginning of process tokens should work fine).

File: pylint/checkers/strings.py [hunk 7]
```diff
@@ -104,7 +105,11 @@
               "Used when a PEP 3101 format string uses a lookup specifier "
               "({a[1]}), but the argument passed for formatting "
               "doesn't contain or doesn't have that key as an attribute.",
-              {'minversion': (2, 7)})
+              {'minversion': (2, 7)}),
+    'W1308': ("Mixing string literals concatenation with comma-separated %s",
```
Comment (rogalski) [inline pylint/checkers/strings.py]: I don't like a notion of mentioning "comma-separated list" / "comma-separated tuples". Are there other lists or tuples in Python?

My attempt at rephrasing would be to do something along lines of:
> "Implicit string concatenation found in {type_name} literal

Same goes to message name.

File: pylint/checkers/strings.py [hunk 8]
```diff
@@ -234,6 +239,22 @@ class StringFormatChecker(BaseChecker):
...
+
+    @check_messages(*(MSGS.keys()))
+    def visit_tuple(self, node):
+        self.check_for_concatenated_strings(node, 'tuple')
+
+    def check_for_concatenated_strings(self, iterable_node, iterable_type):
+        # This check relies on the asumption that StringConstantChecker is run beforehand
+        for elt in iterable_node.elts:
+            if isinstance(elt, astroid.node_classes.Const) and elt.pytype() == 'builtins.str':
+                if StringConstantChecker.string_tokens[(elt.lineno, elt.col_offset)] != elt.value:
```
Comment (rogalski) [inline pylint/checkers/strings.py]: Also, checker classes really should be independent, consider doing your _process tokens_ inside `StringFormatChecker`

File: ChangeLog (hunk 1 above)
Comment (Lucas-C) [reply ChangeLog]: Good idea, fixed

File: ChangeLog (hunk 2 above)
Comment (Lucas-C) [reply ChangeLog]: OK

File: pylint/checkers/strings.py (hunk 3 above)
Comment (Lucas-C) [reply pylint/checkers/strings.py]: I don't see any example of "variablized" detailed message.
Maybe "in a comma-separated iterable definition" instead ?

File: pylint/checkers/strings.py (hunk 4 above)
Comment (Lucas-C) [reply pylint/checkers/strings.py]: Code was already there :)
But ok, boy scout rule applies ;)

File: pylint/checkers/strings.py (hunk 6 above)
Comment (Lucas-C) [reply pylint/checkers/strings.py]: I tested it manually, it is thrown away.
I'm going to try to make this an instance attribute, but it's bit annoying that you precisely recommended a class attribute in one of your earlier comment :(

Do you want to add a unit test ?
If so, could you please then point me to an existing test I could use as an example to perform such "reinitialization" ?

File: pylint/test/functional/str_literals_concatenation_and_comma.py (hunk 5 above)
Comment (Lucas-C) [reply pylint/test/functional/str_literals_concatenation_and_comma.py]: Fixed

File: pylint/checkers/strings.py (hunk 7 above)
Comment (Lucas-C) [reply pylint/checkers/strings.py]: Sounds better, thanks !

File: pylint/checkers/strings.py (hunk 8 above)
Comment (Lucas-C) [reply pylint/checkers/strings.py]: Fixed by introducing `IAstroidAndTokensChecker` interface

**Comment by Lucas-C** [comment]
Tests are failing with Python 2 interpreter.
I'm going to install one on my machine to figure out why (probably either the AST walker or the tokenizer behave differently in regard to string literals concatenation)

**Comment by Lucas-C** [comment]
I fixed this hook for Python 2

**Comment by Lucas-C** [comment]
I don't quite get why the `pylint` tox environment is failing.
There are many `[R]efactor` messages but they all seem to have been preexisting

File: pylint/checkers/strings.py [hunk 9]
```diff
@@ -545,11 +588,11 @@ def process_module(self, module):
         self._unicode_literals = 'unicode_literals' in module.future_imports
 
     def process_tokens(self, tokens):
-        for (tok_type, token, (start_row, _), _, _) in tokens:
+        for (tok_type, token, start, _, _) in tokens:
```
Comment (rogalski) [inline pylint/checkers/strings.py]: this change is redundant

File: pylint/checkers/strings.py [hunk 10]
```diff
@@ -545,11 +588,11 @@ def process_module(self, module):
...
             if tok_type == tokenize.STRING:
                 # 'token' is the whole un-parsed token; we can look at the start
                 # of it to see whether it's a raw or unicode string etc.
-                self.process_string_token(token, start_row)
+                self.process_string_token(token, start[0])
```
Comment (rogalski) [inline pylint/checkers/strings.py]: this change is redundant

File: pylint/checkers/strings.py [hunk 11]
```diff
@@ -527,7 +570,7 @@ class StringConstantChecker(BaseTokenChecker):
                   'String constant might be missing an r or u prefix.',
                   'anomalous-unicode-escape-in-string',
                   'Used when an escape like \\u is encountered in a byte '
-                  'string where it has no effect.'),
+                  'string where it has no effect.')
```
Comment (rogalski) [inline pylint/checkers/strings.py]: this change is redundant

File: pylint/interfaces.py [hunk 12]
```diff
@@ -79,6 +79,18 @@ class IAstroidChecker(IChecker):
     """
 
 
+class IAstroidAndTokensChecker(IAstroidChecker):
```
Comment (rogalski) [inline pylint/interfaces.py]: Isn't `BaseTokenChecker` exact same thing? From what I see `BaseTokenChecker` _is_ a checker, which means it can involve walking ASTs and visiting nodes... Or am I missing something?

File: pylint/lint.py [hunk 13]
```diff
@@ -844,10 +844,12 @@ def _do_check(self, files_or_modules):
                          and c is not self]
         rawcheckers = [c for c in _checkers
                        if interfaces.implements(c, interfaces.IRawChecker)]
+        astcheckers = [c for c in _checkers
```
Comment (rogalski) [inline pylint/lint.py]: If previous comment is true, this won't be needed.

File: pylint/lint.py [hunk 14]
```diff
@@ -844,10 +844,12 @@ def _do_check(self, files_or_modules):
...
+                       if interfaces.implements(c, interfaces.IAstroidChecker)]
         # notify global begin
         for checker in _checkers:
             checker.open()
-            if interfaces.implements(checker, interfaces.IAstroidChecker):
+            if checker in astcheckers:
```
Comment (rogalski) [inline pylint/lint.py]: If previous comment is true, this won't be needed.

File: pylint/lint.py [hunk 15]
```diff
@@ -868,7 +870,7 @@ def _do_check(self, files_or_modules):
             # fix the current file (if the source file was not available or
             # if it's actually a c extension)
             self.current_file = ast_node.file # pylint: disable=maybe-no-member
-            self.check_astroid_module(ast_node, walker, rawcheckers, tokencheckers)
+            self.check_astroid_module(ast_node, walker, rawcheckers, tokencheckers, astcheckers)
```
Comment (rogalski) [inline pylint/lint.py]: If previous comment is true, this won't be needed.

File: pylint/lint.py [hunk 16]
```diff
@@ -922,7 +924,7 @@ def get_ast(self, filepath, modname):
             self.add_message('astroid-error', args=(ex.__class__, ex))
 
     def check_astroid_module(self, ast_node, walker,
-                             rawcheckers, tokencheckers):
+                             rawcheckers, tokencheckers, astcheckers):
```
Comment (rogalski) [inline pylint/lint.py]: If previous comment is true, this won't be needed.

File: pylint/lint.py [hunk 17]
```diff
@@ -946,6 +948,9 @@ def check_astroid_module(self, ast_node, walker,
                 checker.process_module(ast_node)
             for checker in tokencheckers:
                 checker.process_tokens(tokens)
+            for checker in astcheckers:
```
Comment (rogalski) [inline pylint/lint.py]: If previous comment is true, this won't be needed.

File: pylint/test/functional/bad_continuation.py [hunk 18]
```diff
@@ -1,5 +1,5 @@
 """Regression test case for bad-continuation."""
-# pylint: disable=print-statement,using-constant-test
+# pylint: disable=print-statement,implicit-str-concatenation-in-literal-iterable,using-constant-test
```
Comment (rogalski) [inline pylint/test/functional/bad_continuation.py]: Why it's disabled here?

File: tox.ini [hunk 19]
```diff
@@ -7,7 +7,7 @@ deps =
    git+https://github.com/pycqa/astroid@master
    isort
 
-commands = pylint -rn --rcfile={toxinidir}/pylintrc --load-plugins=pylint.extensions.docparams, pylint.extensions.mccabe {envsitepackagesdir}/pylint
+commands = pylint -rn --rcfile={toxinidir}/pylintrc --load-plugins=pylint.extensions.docparams,pylint.extensions.mccabe {envsitepackagesdir}/pylint
```
Comment (rogalski) [inline tox.ini]: if lack of previous warnings was caused by this bug in tox.ini, please revert it for now. We can tackle this problem in another commit.

File: pylint/interfaces.py (hunk 12 above)
Comment (Lucas-C) [reply pylint/interfaces.py]: Yes, well spotted

File: tox.ini (hunk 19 above)
Comment (Lucas-C) [reply tox.ini]: As you want

File: pylint/test/functional/bad_continuation.py (hunk 18 above)
Comment (Lucas-C) [reply pylint/test/functional/bad_continuation.py]: Because of line 37, 58, 60...

File: pylint/interfaces.py (hunk 12 above)
Comment (Lucas-C) [reply pylint/interfaces.py]: There is a side effect however:
making `StringFormatChecker` inherits from `BaseTokenChecker` and implementing `ITokenChecker` forces to change all the `.add_message` call sites in this class, because it now requires a `line` parameter:
https://github.com/PyCQA/pylint/blob/master/pylint/utils.py#L374
https://github.com/PyCQA/pylint/blob/master/pylint/utils.py#L153

**Comment by Lucas-C** [comment]
3 failing tests remain:
- `test_functionality[func_e13xx.py]` : can be resolved by removing the `pprint` string on all the lines in https://github.com/PyCQA/pylint/blob/master/pylint/test/messages/func_e13xx_py30.txt
- `test_functional[string_formatting]` & `test_functional[string_formatting_py3]` : this time there is a missing `object` field in the expected `test_functional.py:OutputLine`

All those failing tests are a consequence of the modified call to `add_message`.
Should I go on and fix all those tests ?

File: ChangeLog [hunk 20]
```diff
@@ -122,6 +122,10 @@ What's New in Pylint 1.8?
       mixing ``Args`` and ``Keyword Args`` in Google docstring.
       Close #1409
 
+    * Check for string concatenation inside lists, sets & tuples, introducing
+      new implicit-str-concatenation-in-literal-iterable (patch by Lucas Cimon)
```
Comment (rogalski) [inline ChangeLog]: Message ID is quite long, maybe we can shorten it up a little bit?

My first thought is something along the lines of `implicit-str-concat-in-sequence`, but I'm open for suggestions.

File: pylint/checkers/strings.py [hunk 21]
```diff
@@ -250,15 +292,15 @@ def visit_binop(self, node):
         except utils.UnsupportedFormatCharacter as e:
             c = format_string[e.index]
             self.add_message('bad-format-character',
-                             node=node, args=(c, ord(c), e.index))
+                             line=node.lineno, args=(c, ord(c), e.index))
```
Comment (rogalski) [inline pylint/checkers/strings.py]: I don't think those modifications are necessary.

File: pylint/test/functional/bad_continuation.py (hunk 18 above)
Comment (rogalski) [reply pylint/test/functional/bad_continuation.py]: I see, that's OK

File: pylint/utils.py [hunk 22]
```diff
@@ -359,7 +359,7 @@ def add_message(self, msg_descr, line=None, node=None, args=None, confidence=UND
 
         If provided, the message string is expanded using args
 
-        AST checkers should must the node argument (but may optionally
+        AST checkers should provide the node argument (but may optionally
```
Comment (rogalski) [inline pylint/utils.py]: I assume that you modified existing `add_message` calls due to this documentation entry. From your case it's clear that 

I'd suggest an entry:

> All emitted messages shall provide at minimum either line in which violation occurred or AST node in which violation occurred. Passing both values is also accepted.

Feel free to rephrase if you have better idea.

File: pylintrc [hunk 23]
```diff
@@ -52,7 +52,8 @@ confidence=
 # no Warning level messages displayed, use"--disable=all --enable=classes
 # --disable=W"
 
-disable=invalid-name,protected-access,fixme,too-many-branches,
+disable=implicit-str-concatenation-in-literal-iterable,
```
Comment (rogalski) [inline pylintrc]: I assume it's disabled here due to message definitions in all checkers, where implicitly joined strings for description are used extensively...

File: ChangeLog (hunk 20 above)
Comment (Lucas-C) [reply ChangeLog]: Ok

File: pylintrc (hunk 23 above)
Comment (Lucas-C) [reply pylintrc]: Yes :)

File: pylint/checkers/strings.py (hunk 21 above)
Comment (Lucas-C) [reply pylint/checkers/strings.py]: They are, sadly:
> making StringFormatChecker inherit from BaseTokenChecker and implement ITokenChecker forces to change all the .add_message call sites in this class, because it now requires a line parameter:
https://github.com/PyCQA/pylint/blob/master/pylint/utils.py#L374
https://github.com/PyCQA/pylint/blob/master/pylint/utils.py#L153

Alternatively, I could move this new check from `StringFormatChecker` into `StringConstantChecker`, which would solve all those issues
I'm going to do so unless someone comment against it here

File: pylint/utils.py (hunk 22 above)
Comment (Lucas-C) [reply pylint/utils.py]: Actually no, it doesn't work like this :(
If the checker inherits `IRawChecker` or `ITokenChecker`, a `line` parameter **must** be passed in `.add_message`, else there **must** be a `node` passed:
https://github.com/PyCQA/pylint/blob/master/pylint/utils.py#L153
https://github.com/PyCQA/pylint/blob/master/pylint/utils.py#L374

File: pylint/checkers/strings.py (hunk 21 above)
Comment (rogalski) [reply pylint/checkers/strings.py]: Extracting your check to individual class is a great idea, go for it!

File: pylint/utils.py (hunk 22 above)
Comment (rogalski) [reply pylint/utils.py]: Unfortunately I did not picked that up when I did review yesterday.

I'm not familiar with this part of the repository, so I cannot give a straightforward answer right now. I'd suggest extracting a new class for your check and wrap everything up so all tests pass (without modifying preexisting test suite). Then we can wait for @PCManticore input on how to tackle this hybrid AST/token checker in terms of emitting messages from them and documenting those checkers in our docs.

**Comment by Lucas-C** [comment]
I think this is ready to be merged

File: doc/development_guide/contribute.rst [hunk 24]
```diff
@@ -166,6 +166,6 @@ A Typical Development Workflow
    It is also possible to give tox a `pytest specifier <https://docs.pytest.org/en/latest/usage.html#specifying-tests-selecting-tests>`_
    to run only your test::
 
-     $ tox pylint/test/test_functional.py::test_functional
+     $ tox -e py36 -- -k 'test_functional[implicit_str_concat_in_sequence]'
```
Comment (PCManticore) [inline doc/development_guide/contribute.rst]: Why did you modify this?

File: pylint/checkers/strings.py [hunk 25]
```diff
@@ -541,15 +549,45 @@ class StringConstantChecker(BaseTokenChecker):
...
     def process_tokens(self, tokens):
-        for (tok_type, token, (start_row, _), _, _) in tokens:
+        for (tok_type, token, start, _, _) in tokens:
             if tok_type == tokenize.STRING:
                 # 'token' is the whole un-parsed token; we can look at the start
                 # of it to see whether it's a raw or unicode string etc.
-                self.process_string_token(token, start_row)
+                self.process_string_token(token, start[0])
+                try:
+                    self.string_tokens[start] = ast.literal_eval(token)
```
Comment (PCManticore) [inline pylint/checkers/strings.py]: What's the purpose of literal_evaling these tokens?

File: pylintrc [hunk 26]
```diff
@@ -52,7 +52,8 @@ confidence=
 # no Warning level messages displayed, use"--disable=all --enable=classes
 # --disable=W"
 
-disable=invalid-name,protected-access,fixme,too-many-branches,
+disable=implicit-str-concat-in-sequence,
```
Comment (PCManticore) [inline pylintrc]: Where exactly do we have this occurring in our code base? It would be nice to see them before merging, since this usually can reveal false positives and whatnot, things we should account for.

File: pylint/checkers/strings.py [hunk 27]
```diff
@@ -541,15 +549,45 @@ class StringConstantChecker(BaseTokenChecker):
...
+    def visit_set(self, node):
+        self.check_for_concatenated_strings(node, 'set')
+
+    @check_messages(*(MSGS.keys()))
+    def visit_tuple(self, node):
+        self.check_for_concatenated_strings(node, 'tuple')
+
+    def check_for_concatenated_strings(self, iterable_node, iterable_type):
+        for elt in iterable_node.elts:
+            if isinstance(elt, Const) and elt.pytype() in _AST_NODE_STR_TYPES:
```
Comment (PCManticore) [inline pylint/checkers/strings.py]: You can probably check the type of `elt.value` instead of using `pytype`

File: doc/development_guide/contribute.rst (hunk 24 above)
Comment (Lucas-C) [reply doc/development_guide/contribute.rst]: To help future contributors by showing more useful CLI flags:
- the ability to execute the tests in only one tox environment
- the ability to execute only one functional test

File: pylint/checkers/strings.py (hunk 25 above)
Comment (Lucas-C) [reply pylint/checkers/strings.py]: To get a string out of a token string.
There are some examples of token values:
- `"'a'"``
- `"u'b'"`
- `"r'b'"`
- `"'b \\\n              c'"`

File: pylint/checkers/strings.py (hunk 27 above)
Comment (Lucas-C) [reply pylint/checkers/strings.py]: I'm not sure to understand how, sorry

The 2nd member of this conditional serves to distinguish between e.g. string constants and integer constants.
Are you suggesting to do something like the collowing code ?
```
if isinstance(elt, Const) and isinstance(elt.value, six.string_types):
```

File: pylintrc (hunk 26 above)
Comment (Lucas-C) [reply pylintrc]: There are 220 violations :(

And yes, this is because of the way message definitions are built.
For example there is the line where the 1st warning is raised:
https://github.com/PyCQA/pylint/blob/master/pylint/extensions/mccabe.py#L131

Interestingly, it is not consistent, and sometimes the newline is escaped:
https://github.com/PyCQA/pylint/blob/master/pylint/checkers/exceptions.py#L62

What do you advise ?
Replace all string concatenations by escaped newlines ? :(

Comment (PCManticore) [reply pylintrc]: I was afraid this is going to happen with this PR, it will emit even for use cases that are valid, such as the ones from pylint's message definition. Given this, I'm finding  the new check too nitpicky. If we were to include this in pylint, then we should restrict it to verify just the implicit string concatenation that occurs on the same line, not on multiple lines.

**Comment by Lucas-C** [comment]
tl;dr : I disagree but I made the required changes

While I totally agree with the checker beeing nickpicky, I think ignoring string concatenations over newlines also restrict the use of this warning.

For example, no warning will be raised for this code:
```
LIST = (
    'item1',
    'item2'
    'item3'
)
```

Whether or not it appears in `pylint` code base, I don't think it is very good code style to have string literal implicit concatenation in comma-separated lists/sets/tuples.
And I think `pylint` should let users decide on the matter by enabling or not this warning, in its most strict form.

**Comment by Lucas-C** [comment]
Are you OK to merge this PR in this form ?

**Comment by Lucas-C** [comment]
I rebased this PR on the master branch. Are you OK to merge it ?

**Comment by PCManticore** [comment]
@Lucas-C Yes, just didn't get to it just yet

**Comment by vicyap** [comment]
I like what @Lucas-C said here: https://github.com/PyCQA/pylint/pull/1655#issuecomment-336118198. It would be useful to also include a check for accidental multiline string concatenations in sequences. Perhaps this could be included but default disabled?

**Comment by Lucas-C** [comment]
I juste rebased this PR so it is not conflicting to merge it.
I don't want to put any pressure on closing this, but maybe the `reviewed-waiting-updates` tag is misleading and should be removed.

File: pylint/checkers/strings.py [hunk 28]
```diff
@@ -554,15 +562,50 @@ class StringConstantChecker(BaseTokenChecker):
...
-        for (tok_type, token, (start_row, _), _, _) in tokens:
+        for i, (tok_type, token, start, _, _) in enumerate(tokens):
             if tok_type == tokenize.STRING:
                 # 'token' is the whole un-parsed token; we can look at the start
                 # of it to see whether it's a raw or unicode string etc.
-                self.process_string_token(token, start_row)
+                self.process_string_token(token, start[0])
+                next_token = tokens[i + 1] if i + 1 < len(tokens) else None
+                try:
+                    self.string_tokens[start] = (ast.literal_eval(token), next_token)
```
Comment (PCManticore) [inline pylint/checkers/strings.py]: I'm not super happy about `literal_eval`-ing every string token, as this might result in a performance hit for big projects that rely on multiple strings in their codebase. Can you find a way to avoid doing it?

Comment (Lucas-C) [reply pylint/checkers/strings.py]: Done

**Comment by Lucas-C** [comment]
This should be good

**Comment by Lucas-C** [comment]
I don't understand the last error in the Travis CI build:
```
pylint/lib/python3.5/site-packages/pylint/checkers/utils.py:893:19: E1124: Argument 'qnames' passed by position and keyword in function call (redundant-keyword-arg)
```

**Comment by rkrzr** [comment]
I'd also be interested in getting this PR merged, since I have gotten bitten by this problem before.

Thanks for the good work @Lucas-C .

**Comment by Lucas-C** [comment]
One year since the first version of this PR... :)
I updated it in order to fix the merge conflicts & failing tests.
Seems good to merge

**Comment by PCManticore** [comment]
@Lucas-C Thank you! Thank you for waiting over an year to get this merged, it's mostly my fault on not getting to it sooner.

**Comment by Lucas-C** [comment]
No need to blame yourself, there's no problem.
Thank you for merging it ! I'm glad it's done :)

**Comment by apnewberry** [comment]
I'm interested in getting this warning when implicitly concatenating across lines. If I'm reading the thread correctly, that was implemented but changed. Is there a simple snippet I can use to get that behavior, perhaps in a plugin?

**Comment by Lucas-C** [comment]
You should get a warning with this code and 2.2.2:
```
t = ('a', 'b' 'c')
```
Example:
```
$ bin/pylint test.py
************* Module test
test.py:1:0: W1403: Implicit string concatenation found in tuple (implicit-str-concat-in-sequence)

--------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 10.00/10, -10.00)
```

**Comment by apnewberry** [comment]
Right, thanks.

```python
"""Pylint 2.2 should warn implicit-str-concat-in-sequence.
https://github.com/PyCQA/pylint/pull/1655

pylint 2.2.2
astroid 2.1.0
Python 3.6.7
[GCC 8.2.0]

Current result:
names.py:13:0: W1403: Implicit string concatenation found in list (implicit-str-concat-in-sequence)

"""

X = ["a", "b" "c", "d"]

Y = ["a",
     "b"
     "c",
     "d"
]

print(len(X))
print(len(Y))

```

I'm interested in getting a warning for the multi-line case, `Y` in the above example. From [this comment](https://github.com/PyCQA/pylint/pull/1655#issuecomment-336118198), I think the warning was added but removed before release. Is there a simple snippet I can use to get that behavior, perhaps in a plugin?

**Comment by Lucas-C** [comment]
Yes, your deductions are totally right.
I was also in favor of getting warnings in the multi-line case.
I do not know of any such option/flag mechanism builtin in pylint, but this is my only contribution to this project so others may be more helpful.
It should probably be doable with a plugin, but I do not know of any existing one that would provide you this behavior. You can try to develop one yourself, I would use it for sure :)

**Comment by PCManticore** [comment]
FWIW, I would be in favour of `implicit-str-concat-in-sequence` to have an option for picking multiline cases as well, as long as it is not the default value.

**Comment by Lucas-C** [comment]
Thanks for your input @PCManticore 
Would you have any suggestion on how such option could be implemented in pylint ? Would that mean another checker ID ?
Also, maybe we should move this discussion to a new issue instead of the bottom of a merged PR :)

**Comment by PCManticore** [comment]
@Lucas-C There's no need for a new check, as long as this one can look for that flag when deciding for multiline vs same line behaviour, that should be enough. Here's some rough instructions:

- Each checker can have an `options` configuration like [this](https://github.com/PyCQA/pylint/blob/master/pylint/checkers/classes.py#L660). That new options should go there.
- Then in the check code itself, you can access any option with `self.config.<name of the option with underscores>`. 
- The new option could then be used both at CLI and in the configuration file.

Let me know if that makes sense, happy to discuss in a separate issue.

**Comment by Lucas-C** [comment]
Thank you for those clear guidelines.
I just sent a PR implementing this.
